src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c

```diff
@@ -0,0 +1,470 @@
+
+    int32_t ret = FAIL;
+
+    int32_t count = (*env)->CallIntMethod(env, ctx->errorList, g_CollectionSize);
+    if (errorsLen < count)
+        goto exit;
+
+    // for (int i = 0; i < erroList.size(); ++i) {
+    //     Throwable error = erroList.get(i);
+    //     << populate errors[i] >>
+    // }
+    for (int32_t i = 0; i < count; ++i)
+    {
+        jobject error = (*env)->CallObjectMethod(env, ctx->errorList, g_ListGet, i);
+        ON_EXCEPTION_PRINT_AND_GOTO(exit);
+        PopulateValidationError(env, error, &errors[i]);
+        (*env)->DeleteLocalRef(env, error);
+    }
+
+    ret = SUCCESS;
+
+exit:
+    return ret;
+}
+
+int32_t AndroidCryptoNative_X509ChainSetCustomTrustStore(X509ChainContext* ctx,
+                                                         jobject* /*X509Certificate*/ customTrustStore,
+                                                         int32_t customTrustStoreLen)
+{
+    assert(ctx != NULL);
+    JNIEnv* env = GetJNIEnv();
+
+    // HashSet<TrustAnchor> anchors = new HashSet<TrustAnchor>(customTrustStoreLen);
+    // for (Certificate cert : customTrustStore) {
+    //     TrustAnchor anchor = new TrustAnchor(cert, null);
+    //     anchors.Add(anchor);
+    // }
+    jobject anchors = (*env)->NewObject(env, g_HashSetClass, g_HashSetCtorWithCapacity, customTrustStoreLen);
+    for (int i = 0; i < customTrustStoreLen; ++i)
+    {
+        jobject anchor = (*env)->NewObject(env, g_TrustAnchorClass, g_TrustAnchorCtor, customTrustStore[i], NULL);
+        (*env)->CallBooleanMethod(env, anchors, g_HashSetAdd, anchor);
+        (*env)->DeleteLocalRef(env, anchor);
+    }
+
+    // params.setTrustAnchors(anchors);
+    (*env)->CallVoidMethod(env, ctx->params, g_PKIXBuilderParametersSetTrustAnchors, anchors);
+
+    (*env)->DeleteLocalRef(env, anchors);
+    return CheckJNIExceptions(env) ? FAIL : SUCCESS;
+}
+
+bool AndroidCryptoNative_X509ChainSupportsRevocationOptions(void)
+{
+    return g_CertPathValidatorGetRevocationChecker != NULL && g_PKIXRevocationCheckerClass != NULL;
+}
+
+static jobject /*HashSet<PKIXRevocationChecker.Option>*/
+GetRevocationCheckerOptions(JNIEnv* env, PAL_X509RevocationMode revocationMode, PAL_X509RevocationFlag revocationFlag)
+{
+    assert(AndroidCryptoNative_X509ChainSupportsRevocationOptions());
+
+    // HashSet<PKIXRevocationChecker.Option> options = new HashSet<PKIXRevocationChecker.Option>(3);
+    jobject options = (*env)->NewObject(env, g_HashSetClass, g_HashSetCtorWithCapacity, 3);
+
+    if (revocationMode == X509RevocationMode_Offline)
+    {
+        // options.add(PKIXRevocationChecker.Option.PREFER_CRLS);
+        jobject preferCrls = (*env)->GetStaticObjectField(
+            env, g_PKIXRevocationCheckerOptionClass, g_PKIXRevocationCheckerOptionPreferCrls);
+        (*env)->CallBooleanMethod(env, options, g_HashSetAdd, preferCrls);
+        (*env)->DeleteLocalRef(env, preferCrls);
+
+        // options.add(PKIXRevocationChecker.Option.NO_FALLBACK);
```

Comment:
I don't know how the revocation works internally here (I haven't found a good description), but I'm not sure that PREFER_CRLS | NO_FALLBACK is quite the same as Offline.

Windows does:

1. Check the OCSP cache for a valid OCSP response.
2. If !offline, do an OCSP fetch
3. Check the CRL cache for a valid CRL entry.
4. If !offline, do a CRL fetch
5. No options remain, report RevocationStatusUnknown

Options 2 and 3 may be reversed, I'm not sure, it's a black box.

On Linux I think we're

1. CRL cache
2. OCSP cache
3. If !offline CRL fetch
4. If !offline OCSP fetch
5. RevocationStatusKnown

but I might be misremembering

macOS is similar, but they have "network is allowed" and "revocation is enabled" separate options.  The former controls both AIA/chain fetch and revocation fetch, so together we can have Offline mode.

If Android's CRL provider is really "you have to have already fetched it for me" but their OCSP provider is live, then I guess it matches, but then we're missing CRL+Online.

Reply:
Ah, boo... Definitely not mapping this properly.

From what I've tested on an Android version where `PKIXRevocationChecker` is available:
- OCSP is live (when enabled)
- Using `PKIXRevocationChecker` (setting options, adding it as a checker), does CRL fetch
- Not using `PKIXRevocationChecker` (just using `setRevocationEnabled`), does not do CRL fetch

Haven't checked an older version yet.

I don't see anything for configuring the fetching. So I guess it would be that if we use `PKIXRevocationChecker`, we are always Online and not using it, we are missing CRL+Online and Offline can't have OCSP.

Reply:
If Offline gets treated as Online, that's probably not the end of the world.  Where do the CRLs come from in "Not using PKIXRevocationChecker"?  If it's from an internal cache, then I think the mapping is

* Online: OCSP enabled, use PKIXRevocationChecker
* Offline: OCSP disabled, just setRevocationEnabled(true)
* NoCheck: setRevocationEnabled(false)

If it's "you have to give it to us" then perhaps just do Offline => Online for now with a vWhenever issue for properly respecting Offline mode.  (Presumably you can get the CRLs off of the chain to save for later, and then we'd just have to squirrel them away somewhere...)

Reply:
Looks like checking an internal cache is wrapped up in using `PKIXRevocationChecker` and when not using that, it is 'you have to give it to us'.

Reply:
Treating `Offline` as `Online` now (with a debug log message).